Saving a config results in an error.

**Describe the bug**
When I save a config, it results in an error.  I'll copy the payload below and the error message.

**To Reproduce**
Steps to reproduce the behavior:
Here is my payload:
{
    "configSensors": [
        {
            "dataTypes": {
                "SHEF-PE": "HFIRR",
                "CWMS": "Elev-Forebay"
            },
            "properties": {
                "CwmsParamType": "Inst",
                "CwmsInterval": "0",
                "CwmsVersion": "IRIDIUM-RAW",
                "CwmsDuration": "0"
            },
            "recordingInterval": 300,
            "recordingMode": "F",
            "sensorName": "HF",
            "sensorNumber": 1,
            "timeOfFirstSample": 0,
            "usgsStatCode": "",
            "absoluteMax": 4000,
            "absoluteMin": -15
        },
        {
            "dataTypes": {
                "CWMS": "Stor"
            },
            "properties": {
                "CwmsParamType": "Inst",
                "CwmsInterval": "0",
                "CwmsVersion": "IRIDIUM-RAW",
                "CwmsDuration": "0"
            },
            "recordingInterval": 300,
            "recordingMode": "F",
            "sensorName": "Storage",
            "sensorNumber": 2,
            "timeOfFirstSample": 0,
            "usgsStatCode": "",
            "absoluteMax": 1.7976931348623157e+308,
            "absoluteMin": 1.7976931348623157e+308
        }
    ],
    "description": "",
    "name": "APP-IR",
    "numPlatforms": "0",
    "scripts": [
        {
            "name": "ir",
            "headerType": null,
            "scriptSensors": [
                {
                    "sensorNumber": 1,
                    "unitConverter": {
                        "ucId": -1,
                        "fromAbbr": "raw",
                        "toAbbr": "ft",
                        "algorithm": "usgs-standard",
                        "a": 0.01,
                        "b": 0,
                        "c": 1,
                        "d": 1800,
                        "e": 0,
                        "f": 0
                    }
                }
            ],
            "formatStatements": [],
            "dataOrder": "U"
        }
    ],
    "configId": 2
}
Now, when I post this to my database, I get the following error message - (It's a foreign key constraint, so it could possibly be specific to this DB instance).
07-Apr-2025 14:01:06.308 WARNING [http-nio-8080-exec-3] org.opendcs.odcsapi.res.AppExceptionMapper.handle There was an error.  Please contact your sys admin.
	org.opendcs.odcsapi.dao.DbException: Error saving config
		at org.opendcs.odcsapi.res.ConfigResources.postConfig(ConfigResources.java:315)
		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:498)
		at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
		at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:146)
		at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:189)
		at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176)
		at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:93)
		at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:478)
		at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:400)
		at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:81)
		at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:256)
		at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
		at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
		at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
		at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
		at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
		at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
		at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:235)
		at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:684)
		at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:394)
		at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346)
		at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:358)
		at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:311)
		at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205)
		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:212)
		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:156)
		at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51)
		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:181)
		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:156)
		at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:168)
		at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90)
		at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:483)
		at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:130)
		at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93)
		at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:679)
		at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
		at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
		at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:617)
		at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63)
		at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:932)
		at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1694)
		at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52)
		at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
		at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
		at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
		at java.lang.Thread.run(Thread.java:748)
	Caused by: decodes.db.DatabaseException: writeConfig: 
		at decodes.sql.SqlDatabaseIO.writeConfig(SqlDatabaseIO.java:1423)
		at org.opendcs.odcsapi.res.ConfigResources.postConfig(ConfigResources.java:308)
		... 48 more
	Caused by: org.postgresql.util.PSQLException: ERROR: insert or update on table "configsensordatatype" violates foreign key constraint "configsensordatatype_datatypeid_fkey"
  Detail: Key (datatypeid)=(479) is not present in table "datatype".
		at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2733)
		at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2420)
		at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:372)
		at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:517)
		at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:434)
		at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:356)
		at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:341)
		at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:317)
		at org.postgresql.jdbc.PgStatement.executeUpdate(PgStatement.java:290)
		at org.apache.tomcat.dbcp.dbcp2.DelegatingStatement.executeUpdate(DelegatingStatement.java:264)
		at org.apache.tomcat.dbcp.dbcp2.DelegatingStatement.executeUpdate(DelegatingStatement.java:264)
		at decodes.sql.SqlDbObjIo.executeUpdate(SqlDbObjIo.java:304)
		at decodes.sql.ConfigListIO.insert(ConfigListIO.java:673)
		at decodes.sql.ConfigListIO.insertConfigSensors(ConfigListIO.java:576)
		at decodes.sql.ConfigListIO.update(ConfigListIO.java:469)
		at decodes.sql.ConfigListIO.write(ConfigListIO.java:414)
		at decodes.sql.SqlDatabaseIO.writeConfig(SqlDatabaseIO.java:1414)
		... 49 more


**Expected behavior**
The expected behavior would be that the config would be saved.

**Additional context**
This is a foreign key constraint issue, so if you want to debug it, feel free to contact me and we can take a look on my DB.  It could be an issue that has existed, it could even be a usability issue that has existed for a while.

--- a/opendcs-rest-api/src/main/java/org/opendcs/odcsapi/res/ConfigResources.java
+++ b/opendcs-rest-api/src/main/java/org/opendcs/odcsapi/res/ConfigResources.java
@@ -23,6 +23,7 @@ import java.util.Map;
 import java.util.Vector;
 import javax.annotation.security.RolesAllowed;
 
+import decodes.db.DataTypeSet;
 import decodes.db.FormatStatement;
 import io.swagger.v3.oas.annotations.Operation;
 import io.swagger.v3.oas.annotations.Parameter;
@@ -335,7 +336,9 @@ public final class ConfigResources extends OpenDcsResource
 		DatabaseIO dbIo = getLegacyDatabase();
 		try
 		{
-			PlatformConfig pc = map(config);
+			DataTypeSet dataTypeSet = new DataTypeSet();
+			dbIo.readDataTypeSet(dataTypeSet);
+			PlatformConfig pc = map(config, dataTypeSet);
 			dbIo.writeConfig(pc);
 			return Response.status(HttpServletResponse.SC_CREATED)
 					.entity(map(pc))
@@ -351,7 +354,7 @@ public final class ConfigResources extends OpenDcsResource
 		}
 	}
 
-	static PlatformConfig map(ApiPlatformConfig config) throws DbException
+	static PlatformConfig map(ApiPlatformConfig config, DataTypeSet dataTypeSet) throws DbException
 	{
 		try
 		{
@@ -382,7 +385,11 @@ public final class ConfigResources extends OpenDcsResource
 				configSensor.setUsgsStatCode(sensor.getUsgsStatCode());
 				for (Map.Entry<String, String> entry : sensor.getDataTypes().entrySet())
 				{
-					DataType dt = new DataType(entry.getKey(), entry.getValue());
+					DataType dt = dataTypeSet.get(entry.getKey(), entry.getValue());
+					if(dt == null )
+					{
+						dt = new DataType(entry.getKey(), entry.getValue());
+					}
 					configSensor.addDataType(dt);
 				}
 				for (String name : sensor.getProperties().stringPropertyNames())

--- a/opendcs-rest-api/src/test/java/org/opendcs/odcsapi/res/ConfigResourcesTest.java
+++ b/opendcs-rest-api/src/test/java/org/opendcs/odcsapi/res/ConfigResourcesTest.java
@@ -9,6 +9,8 @@ import java.util.Properties;
 import java.util.Vector;
 
 import decodes.db.ConfigSensor;
+import decodes.db.DataType;
+import decodes.db.DataTypeSet;
 import decodes.db.DecodesScript;
 import decodes.db.FormatStatement;
 import decodes.db.PlatformConfig;
@@ -137,7 +139,11 @@ final class ConfigResourcesTest
 		configSensors.add(configSensor);
 		apiConfig.setConfigSensors(configSensors);
 
-		PlatformConfig config = map(apiConfig);
+		DataTypeSet dataTypeSet = new DataTypeSet();
+		DataType dataType = new DataType("Test data type", "Test data type description");
+		dataType.setId(DbKey.createDbKey(1234L));
+		dataTypeSet.add(dataType);
+		PlatformConfig config = map(apiConfig, dataTypeSet);
 
 		assertNotNull(config);
 		assertEquals(apiConfig.getNumPlatforms(), config.numPlatformsUsing);
@@ -168,6 +174,7 @@ final class ConfigResourcesTest
 			{
 				assertEquals(sensorConfig.getDataType().getStandard(), entry.getKey());
 				assertEquals(sensorConfig.getDataType().getCode(), entry.getValue());
+				assertEquals(dataType.getId(), sensorConfig.getDataType().getId());
 			}
 			assertEquals(apiConfig.getConfigSensors().get(0).getProperties(), sensorConfig.getProperties());
 		}

